src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java

```diff
@@ -472,47 +468,192 @@ public synchronized void write(String fileName) throws IOException {
+            List<Field> staticFields = new ArrayList<>();
+            List<Field> instanceFields = new ArrayList<>();
+            Iterator<Field> itr = null;
+            // loader + signer + protectionDomain + 2 reserved + fieldSize + cpool entris number
+            size += OBJ_ID_SIZE * 5 + INT_SIZE + SHORT_SIZE;
+            for (itr = declaredFields.iterator(); itr.hasNext();) {
+                Field field = itr.next();
+                if (field.isStatic()) {
+                    staticFields.add(field);
+                } else {
+                    instanceFields.add(field);
+                }
+            }
+            // size of static field descriptors
+            size += calculateFieldDescriptorsDumpRecordSize(staticFields, ik);
+            // size of instance field descriptors
+            size += calculateFieldDescriptorsDumpRecordSize(instanceFields, null);
+        } else {
+            size += OBJ_ID_SIZE * 5  + INT_SIZE + SHORT_SIZE * 3;
+        }
+        return size;
+    }
+
+    private int calculateFieldDescriptorsDumpRecordSize(List<Field> fields, InstanceKlass ik)
+            throws IOException {
+        int size = 0;
+        size += SHORT_SIZE;
+        for (Iterator<Field> itr = fields.iterator(); itr.hasNext();) {
+            Field field = itr.next();
+            char typeCode = (char) field.getSignature().getByteAt(0);
+            size += OBJ_ID_SIZE + BYTE_SIZE;
+            // ik == null for instance fields
+            if (ik != null) {
+                // static field
+                size += calculateFieldDumpRecordSize(field, ik.getJavaMirror());
+            }
         }
+        return size;
+    }
+
+    private int calculateFieldDumpRecordSize(Field field, Oop oop) throws IOException {
+        char typeCode = (char) field.getSignature().getByteAt(0);
+        int size = 0;
+        switch (typeCode) {
+            case JVM_SIGNATURE_BOOLEAN:
+            case JVM_SIGNATURE_BYTE:
+                size++;
+                break;
+            case JVM_SIGNATURE_CHAR:
+            case JVM_SIGNATURE_SHORT:
+                size += 2;
+                break;
+            case JVM_SIGNATURE_INT:
+            case JVM_SIGNATURE_FLOAT:
+                size += 4;
+                break;
+            case JVM_SIGNATURE_CLASS:
+            case JVM_SIGNATURE_ARRAY:
+                size += OBJ_ID_SIZE;
+                break;
+            case JVM_SIGNATURE_LONG:
+            case JVM_SIGNATURE_DOUBLE:
+                size += 8;
+                break;
+            default:
+                throw new RuntimeException("should not reach here");
+        }
+        return size;
+    }
+
+    private int calculateClassInstanceDumpRecordSize(Instance instance) {
+        Klass reflectedKlass = java_lang_Class.asKlass(instance);
+        // dump instance record only for primitive type Class objects.
+        // all other Class objects are covered by writeClassDumpRecords.
```

Comment:
Formatting nit: If the comment is ended with dot then start it from capital a letter.